File: LISTA CIRCULAR/nodo.java
const readline = require('readline');

class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class CircularLinkedList {
    constructor() {
        this.head = null;
        this.tail = null;
        this.rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
    }

    clearScreen() {
        console.clear();
    }

    async beginset() {
        this.clearScreen();
        console.log("=== INSERTAR AL INICIO ===");
        const item = await this.question("Ingrese el valor del nodo: ");
        
        const new_node = new Node(parseInt(item));
        
        if (this.head === null) {
            this.head = new_node;
            this.tail = new_node;
            new_node.next = this.head;
            console.log("Primer nodo insertado al inicio (lista circular creada)");
        } else {
            new_node.next = this.head;
            this.head = new_node;
            this.tail.next = this.head;
            console.log("Nodo insertado al inicio correctamente");
        }
    }

    async lastinsert() {
        this.clearScreen();
        console.log("=== INSERTAR AL FINAL ===");
        const item = await this.question("Ingrese el valor del nodo: ");
        
        const new_node = new Node(parseInt(item));
        
        if (this.head === null) {
            this.head = new_node;
            this.tail = new_node;
            new_node.next = this.head;
            console.log("Primer nodo insertado al final (lista circular creada)");
        } else {
            this.tail.next = new_node;
            this.tail = new_node;
            new_node.next = this.head;
            console.log("Nodo insertado al final correctamente");
        }
    }

    async randominsert() {
        this.clearScreen();
        console.log("=== INSERTAR EN POSICION ESPECIFICA ===");
        const item = await this.question("Ingrese el valor del elemento: ");
        const loc = await this.question("Ingrese la posicion despues de la cual desea insertar: ");
        
        if (this.head === null) {
            console.log("La lista esta vacia. Insertando como primer nodo...");
            const new_node = new Node(parseInt(item));
            this.head = new_node;
            this.tail = new_node;
            new_node.next = this.head;
            return;
        }
        
        if (loc < 0) {
            console.log("Posicion no valida");
            return;
        }
        
        const new_node = new Node(parseInt(item));
        let temp = this.head;
        let count = 0;
        
        while (count < loc) {
            temp = temp.next;
            count++;
            if (temp === this.head && loc > 0) {
                console.log("Posicion fuera de rango");
                return;
            }
        }
        
        new_node.next = temp.next;
        temp.next = new_node;
        
        if (temp === this.tail) {
            this.tail = new_node;
        }
        
        console.log(Nodo insertado correctamente despues de la posicion ${loc});
    }

    begin_delete() {
        this.clearScreen();
        console.log("=== ELIMINAR DEL INICIO ===");
        if (this.head === null) {
            console.log("La lista esta vacia");
        } else if (this.head === this.tail) {
            this.head = null;
            this.tail = null;
            console.log("Unico nodo eliminado");
        } else {
            this.head = this.head.next;
            this.tail.next = this.head;
            console.log("Nodo eliminado del inicio correctamente");
        }
    }

    last_delete() {
        this.clearScreen();
        console.log("=== ELIMINAR DEL FINAL ===");
        if (this.head === null) {
            console.log("La lista esta vacia");
        } else if (this.head === this.tail) {
            this.head = null;
            this.tail = null;
            console.log("Unico nodo eliminado");
        } else {
            let temp = this.head;
            while (temp.next !== this.tail) {
                temp = temp.next;
            }
            this.tail = temp;
            this.tail.next = this.head;
            console.log("Nodo eliminado del final correctamente");
        }
    }

    async random_delete() {
        this.clearScreen();
        console.log("=== ELIMINAR NODO EN POSICION ESPECIFICA ===");
        const loc = await this.question("Ingrese la posicion del nodo que desea eliminar (0-based): ");
        
        if (this.head === null) {
            console.log("La lista esta vacia");
            return;
        }
        
        if (parseInt(loc) === 0) {
            this.begin_delete();
            return;
        }
        
        let temp = this.head;
        let prev = null;
        let count = 0;
        
        while (count < parseInt(loc)) {
            prev = temp;
            temp = temp.next;
            count++;
            if (temp === this.head) {
                console.log("Posicion fuera de rango");
                return;
            }
        }
        
        prev.next = temp.next;
        
        if (temp === this.tail) {
            this.tail = prev;
        }
        
        console.log(Nodo eliminado de la posicion ${loc});
    }

    async search() {
        this.clearScreen();
        console.log("=== BUSCAR ELEMENTO ===");
        if (this.head === null) {
            console.log("Lista vacia");
            return;
        }
        
        const item = await this.question("Ingrese el elemento que desea buscar: ");
        
        let temp = this.head;
        let i = 0;
        let found = false;
        
        do {
            if (temp.data === parseInt(item)) {
                console.log(Elemento encontrado en la posicion ${i});
                found = true;
                break;
            }
            temp = temp.next;
            i++;
        } while (temp !== this.head);
        
        if (!found) {
            console.log("Elemento no encontrado");
        }
    }

    display() {
        this.clearScreen();
        console.log("=== MOSTRAR LISTA CIRCULAR ===");
        if (this.head === null) {
            console.log("La lista esta vacia");
        } else {
            let temp = this.head;
            process.stdout.write("Elementos de la lista: ");
            do {
                process.stdout.write(temp.data + " ");
                temp = temp.next;
            } while (temp !== this.head);
            console.log();
            console.log(Head: ${this.head.data}, Tail: ${this.tail.data}, Tail->next: ${this.tail.next.data});
            console.log("(La lista es circular - el ultimo nodo apunta al primero)");
        }
    }

    question(prompt) {
        return new Promise((resolve) => {
            this.rl.question(prompt, resolve);
        });
    }

    async menu() {
        let choice;
        do {
            this.clearScreen();
            console.log("\n\n********* MENU LISTA CIRCULAR *********");
            console.log("\nElija una opcion del siguiente menu:");
            console.log("1. Insertar al inicio");
            console.log("2. Insertar al final");
            console.log("3. Insertar en posicion especifica");
            console.log("4. Eliminar del inicio");
            console.log("5. Eliminar del final");
            console.log("6. Eliminar nodo en posicion especifica");
            console.log("7. Buscar elemento");
            console.log("8. Mostrar lista");
            console.log("9. Salir");
            
            choice = await this.question("\nIngrese su eleccion: ");
            
            switch (choice) {
                case '1': await this.beginset(); break;
                case '2': await this.lastinsert(); break;
                case '3': await this.randominsert(); break;
                case '4': this.begin_delete(); break;
                case '5': this.last_delete(); break;
                case '6': await this.random_delete(); break;
                case '7': await this.search(); break;
                case '8': this.display(); break;
                case '9': console.log("Saliendo del programa..."); break;
                default: console.log("Por favor ingrese una opcion valida..."); break;
            }
            
            if (choice !== '9') {
                console.log("\n\nPresione Enter para continuar...");
                await this.question("");
            }
        } while (choice !== '9');
        
        this.rl.close();
    }
}

// Ejecutar el programa
const lista = new CircularLinkedList();
lista.menu();

File: LISTA CIRCULAR/nodo.cs
using System;

class Node {
    public int data;
    public Node next;
    
    public Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class CircularLinkedList {
    private Node head;
    private Node tail;
    
    private void ClearScreen() {
        Console.Clear();
    }
    
    public void Beginset() {
        ClearScreen();
        Console.WriteLine("=== INSERTAR AL INICIO ===");
        Console.Write("Ingrese el valor del nodo: ");
        int item = Convert.ToInt32(Console.ReadLine());
        
        Node new_node = new Node(item);
        
        if (head == null) {
            head = new_node;
            tail = new_node;
            new_node.next = head;
            Console.WriteLine("Primer nodo insertado al inicio (lista circular creada)");
        } else {
            new_node.next = head;
            head = new_node;
            tail.next = head;
            Console.WriteLine("Nodo insertado al inicio correctamente");
        }
    }
    
    public void Lastinsert() {
        ClearScreen();
        Console.WriteLine("=== INSERTAR AL FINAL ===");
        Console.Write("Ingrese el valor del nodo: ");
        int item = Convert.ToInt32(Console.ReadLine());
        
        Node new_node = new Node(item);
        
        if (head == null) {
            head = new_node;
            tail = new_node;
            new_node.next = head;
            Console.WriteLine("Primer nodo insertado al final (lista circular creada)");
        } else {
            tail.next = new_node;
            tail = new_node;
            new_node.next = head;
            Console.WriteLine("Nodo insertado al final correctamente");
        }
    }
    
    public void Randominsert() {
        ClearScreen();
        Console.WriteLine("=== INSERTAR EN POSICION ESPECIFICA ===");
        Console.Write("Ingrese el valor del elemento: ");
        int item = Convert.ToInt32(Console.ReadLine());
        Console.Write("Ingrese la posicion despues de la cual desea insertar: ");
        int loc = Convert.ToInt32(Console.ReadLine());
        
        if (head == null) {
            Console.WriteLine("La lista esta vacia. Insertando como primer nodo...");
            Node new_node = new Node(item);
            head = new_node;
            tail = new_node;
            new_node.next = head;
            return;
        }
        
        if (loc < 0) {
            Console.WriteLine("Posicion no valida");
            return;
        }
        
        Node new_node2 = new Node(item);
        Node temp = head;
        int count = 0;
        
        while (count < loc) {
            temp = temp.next;
            count++;
            if (temp == head && loc > 0) {
                Console.WriteLine("Posicion fuera de rango");
                return;
            }
        }
        
        new_node2.next = temp.next;
        temp.next = new_node2;
        
        if (temp == tail) {
            tail = new_node2;
        }
        
        Console.WriteLine($"Nodo insertado correctamente despues de la posicion {loc}");
    }
    
    public void Begin_delete() {
        ClearScreen();
        Console.WriteLine("=== ELIMINAR DEL INICIO ===");
        if (head == null) {
            Console.WriteLine("La lista esta vacia");
        } else if (head == tail) {
            head = null;
            tail = null;
            Console.WriteLine("Unico nodo eliminado");
        } else {
            head = head.next;
            tail.next = head;
            Console.WriteLine("Nodo eliminado del inicio correctamente");
        }
    }
    
    public void Last_delete() {
        ClearScreen();
        Console.WriteLine("=== ELIMINAR DEL FINAL ===");
        if (head == null) {
            Console.WriteLine("La lista esta vacia");
        } else if (head == tail) {
            head = null;
            tail = null;
            Console.WriteLine("Unico nodo eliminado");
        } else {
            Node temp = head;
            while (temp.next != tail) {
                temp = temp.next;
            }
            tail = temp;
            tail.next = head;
            Console.WriteLine("Nodo eliminado del final correctamente");
        }
    }
    
    public void Random_delete() {
        ClearScreen();
        Console.WriteLine("=== ELIMINAR NODO EN POSICION ESPECIFICA ===");
        Console.Write("Ingrese la posicion del nodo que desea eliminar (0-based): ");
        int loc = Convert.ToInt32(Console.ReadLine());
        
        if (head == null) {
            Console.WriteLine("La lista esta vacia");
            return;
        }
        
        if (loc == 0) {
            Begin_delete();
            return;
        }
        
        Node temp = head;
        Node prev = null;
        int count = 0;
        
        while (count < loc) {
            prev = temp;
            temp = temp.next;
            count++;
            if (temp == head) {
                Console.WriteLine("Posicion fuera de rango");
                return;
            }
        }
        
        prev.next = temp.next;
        
        if (temp == tail) {
            tail = prev;
        }
        
        Console.WriteLine($"Nodo eliminado de la posicion {loc}");
    }
    
    public void Search() {
        ClearScreen();
        Console.WriteLine("=== BUSCAR ELEMENTO ===");
        if (head == null) {
            Console.WriteLine("Lista vacia");
            return;
        }
        
        Console.Write("Ingrese el elemento que desea buscar: ");
        int item = Convert.ToInt32(Console.ReadLine());
        
        Node temp = head;
        int i = 0;
        bool found = false;
        
        do {
            if (temp.data == item) {
                Console.WriteLine($"Elemento encontrado en la posicion {i}");
                found = true;
                break;
            }
            temp = temp.next;
            i++;
        } while (temp != head);
        
        if (!found) {
            Console.WriteLine("Elemento no encontrado");
        }
    }
    
    public void Display() {
        ClearScreen();
        Console.WriteLine("=== MOSTRAR LISTA CIRCULAR ===");
        if (head == null) {
            Console.WriteLine("La lista esta vacia");
        } else {
            Node temp = head;
            Console.Write("Elementos de la lista: ");
            do {
                Console.Write(temp.data + " ");
                temp = temp.next;
            } while (temp != head);
            Console.WriteLine();
            Console.WriteLine($"Head: {head.data}, Tail: {tail.data}, Tail->next: {tail.next.data}");
            Console.WriteLine("(La lista es circular - el ultimo nodo apunta al primero)");
        }
    }
    
    public void Menu() {
        int choice;
        do {
            ClearScreen();
            Console.WriteLine("\n\n********* MENU LISTA CIRCULAR *********");
            Console.WriteLine("\nElija una opcion del siguiente menu:");
            Console.WriteLine("1. Insertar al inicio");
            Console.WriteLine("2. Insertar al final");
            Console.WriteLine("3. Insertar en posicion especifica");
            Console.WriteLine("4. Eliminar del inicio");
            Console.WriteLine("5. Eliminar del final");
            Console.WriteLine("6. Eliminar nodo en posicion especifica");
            Console.WriteLine("7. Buscar elemento");
            Console.WriteLine("8. Mostrar lista");
            Console.WriteLine("9. Salir");
            
            Console.Write("\nIngrese su eleccion: ");
            choice = Convert.ToInt32(Console.ReadLine());
            
            switch (choice) {
                case 1: Beginset(); break;
                case 2: Lastinsert(); break;
                case 3: Randominsert(); break;
                case 4: Begin_delete(); break;
                case 5: Last_delete(); break;
                case 6: Random_delete(); break;
                case 7: Search(); break;
                case 8: Display(); break;
                case 9: Console.WriteLine("Saliendo del programa..."); break;
                default: Console.WriteLine("Por favor ingrese una opcion valida..."); break;
            }
            
            if (choice != 9) {
                Console.WriteLine("\n\nPresione Enter para continuar...");
                Console.ReadLine();
            }
        } while (choice != 9);
    }
}

class Program {
    static void Main(string[] args) {
        CircularLinkedList lista = new CircularLinkedList();
        lista.Menu();
    }
}